### PR TITLE
Create CoreGamePlay

### PR DESCRIPTION
**Pull request by SwapnilOutscal**
- Added PathSystem
- Added PlayerSystem
- Added InputSystem
- Added EnemySystem (Basic)
- Added Interactable Pickup System (Basic)
- Added 3 Levels

### CODE REVIEW

**Review by mayankgrover — COMMENTED** [review]
why is there a apk file in the repository?

File: hitman-go/Assets/Scripts/Common/NodeData.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                }
+            }
+            return false;
+        }
+    }
+    [Serializable]
+    public class ScriptableNode
+    {
+        public NodeData node;
+        public int up=-1, down=-1, left=-1, right=-1;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Common/NodeData.cs]: why 4 variables with same value?

File: hitman-go/Assets/Scripts/Common/NodeData.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+using UnityEngine;
+using System;
+using System.Collections.Generic;
+
+namespace Common
+{
+    public enum NodeProperty
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Common/NodeData.cs]: separate files

File: hitman-go/Assets/Scripts/Common/GameplayEnums.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
+namespace Common
+{
+    public enum Directions
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Common/GameplayEnums.cs:3]: separate files

File: hitman-go/Assets/Scripts/Enemy/CircularCopEnemyController.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+using GameState.Interface;
+using System.Collections;
+
+namespace Enemy
+{
+    public class CircularCopEnemyController : EnemyController
+    {
+
+
+        public CircularCopEnemyController(IEnemyService _enemyService, IPathService _pathService, IGameService _gameService, Vector3 _spawnLocation, EnemyScriptableObject _enemyScriptableObject, int currentNodeID, Directions spawnDirection) : base(_enemyService, _pathService, _gameService, _spawnLocation, _enemyScriptableObject, currentNodeID, spawnDirection)
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/CircularCopEnemyController.cs]: too many dependencies - discuss

File: hitman-go/Assets/Scripts/Enemy/EnemyController.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            gameService = _gameService;
+            SpawnEnemyView();
+        }
+    
+        protected virtual void SpawnEnemyView()
+        {
+            //SPAWN ENEMY VIEW
+            enemyInstance=GameObject.Instantiate(enemyScriptableObject.enemyPrefab.gameObject);
+            currentEnemyView = enemyInstance.GetComponent<IEnemyView>();
+            enemyInstance.transform.localPosition = spawnLocation;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/EnemyController.cs]: anything to do with the view should be inside view script not directly in controller

File: hitman-go/Assets/Scripts/Common/GameInstaller.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        Container.Bind<IPathService>()
+            .To<PathService>()
+            .AsSingle()
+            .NonLazy();
+
+        Container.BindInterfacesAndSelfTo<GameService>()
+            .AsSingle()
+            .NonLazy();
+
+        Container.BindSignal<StateChangeSignal>().ToMethod<GameService>(x=>x.ChangeState).FromResolve();
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Common/GameInstaller.cs]: why are we binding to the signal here and not subscribing like other services

File: hitman-go/Assets/Scripts/Enemy/EnemyService.cs [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        {
+            if(gameService.GetCurrentState()==GameStatesType.ENEMYSTATE)
+            {
+                PerformMovement();
+            }
+        }
+
+        private void PerformMovement()
+        {
+            List<int> enemyKey = new List<int>();
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/EnemyService.cs]: why do we need another list?

File: hitman-go/Assets/Scripts/Enemy/EnemyService.cs [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+            }
+        }
+
+        public void TriggerPlayerDeath()
+        {
+           
+            signalBus.TryFire(new PlayerDeathSignal());
+        }
+
+        private void SpawnSingleEnemy(EnemyScriptableObject _enemyScriptableObject)
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/EnemyService.cs]: too big needs refactoring asap

File: hitman-go/Assets/Scripts/Enemy/PatrollingEnemyController.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+        public PatrollingEnemyController(IEnemyService _enemyService, IPathService _pathService, IGameService _gameService, Vector3 _spawnLocation, EnemyScriptableObject _enemyScriptableObject, int currentNodeID, Directions spawnDirection) : base(_enemyService, _pathService, _gameService, _spawnLocation, _enemyScriptableObject, currentNodeID, spawnDirection)
+        {
+
+
+        }
+
+        protected override void MoveToNextNode(int nodeID)
+        {
+            currentEnemyView.GetGameObject().transform.localPosition = pathService.GetNodeLocation(nodeID);
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/PatrollingEnemyController.cs]: move to view

File: hitman-go/Assets/Scripts/Enemy/IEnemyController.cs [hunk 10]
```diff
@@ -0,0 +1,10 @@
+﻿using UnityEngine;
+using System.Collections;
+
+namespace Enemy
+{
+    public interface IEnemyController
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/IEnemyController.cs]: why is it empty? no one is calling any function on enemy controlller?

File: hitman-go/Assets/Scripts/PathSystem/PathService.cs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace PathSystem
+{
+    public class PathService : IPathService
+    {
+        List<int> shortestPath;
+        GameObject line;
+        NodeControllerView nodeprefab, targetNode;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/PathSystem/PathService.cs]: why is service holding view references ?

File: hitman-go/Assets/Scripts/PathSystem/PathService.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
+using Common;
+using PathSystem.NodesScript;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace PathSystem
+{
+    public class PathService : IPathService
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/PathSystem/PathService.cs:8]: getting too big - needs refactoring

File: hitman-go/Assets/Scripts/Player/Interfaces/IPlayerController.cs [hunk 13]
```diff
@@ -0,0 +1,10 @@
+﻿using UnityEngine;
+using System.Collections;
+
+namespace Player
+{
+    public interface IPlayerController
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Player/Interfaces/IPlayerController.cs:9]: empty why ?

File: hitman-go/Assets/Scripts/Player/PlayerController.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+using UnityEngine;
+
+namespace Player
+{
+    public class PlayerController : IPlayerController
+    {
+        private IPlayerService currentPlayerService;
+        private IPlayerView currentPlayerView;
+        private PlayerScriptableObject scriptableObject;
+        GameObject playerInstance;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Player/PlayerController.cs:24]: controller should not do anything on the player instance. it should only be calling player view functions

File: hitman-go/Assets/Scripts/Player/PlayerService.cs [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        private PlayerDeathSignal playerDeathSignal;        
+        private PlayerScriptableObject playerScriptableObject;
+        private Vector3 spawnLocation;
+        private IGameService gameService;
+        private bool isPlayerDead=false;
+        private int playerNodeID;
+
+        public PlayerService(IPathService _pathService,IGameService  _gameService,PlayerScriptableObject _playerScriptableObject, SignalBus signalBus)
+        {
+            _signalBus = signalBus;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Player/PlayerService.cs]: inconsistent naming with "_"

File: hitman-go/Assets/Scripts/Player/PlayerService.cs [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+            }
+            Vector3 nextLocation = currentPathService.GetNodeLocation(nextNodeID);
+        
+            playerController.MoveToLocation(nextLocation);
+            playerNodeID = nextNodeID;           
+            if (CheckForFinishCondition())
+            {
+                Debug.Log("Game finished");
+            }
+            _signalBus.TryFire(new StateChangeSignal());
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Player/PlayerService.cs]: shouldn't we fire game finished signal when the game is over ?

File: hitman-go/Assets/Scripts/Player/PlayerView.cs [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+namespace Player
+{
+    public class PlayerView : MonoBehaviour, IPlayerView
+    {
+        public void DisablePlayer()
+        {
+            gameObject.SetActive(false);
+        }
+
+        public GameObject GetGameObject()
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Player/PlayerView.cs]: remove

**Review by mayankgrover — CHANGES_REQUESTED** [review]

File: hitman-go/Assets/Scripts/Player/PlayerService.cs (hunk 16 above)
Comment (VinayKashyap15) [reply hitman-go/Assets/Scripts/Player/PlayerService.cs]: there is another game over signal  for that.
 Game finished means that finish condition is met

File: hitman-go/Assets/Scripts/Player/PlayerView.cs (hunk 17 above)
Comment (VinayKashyap15) [reply hitman-go/Assets/Scripts/Player/PlayerView.cs]: okay

File: hitman-go/Assets/Scripts/Enemy/EnemyService.cs (hunk 7 above)
Comment (VinayKashyap15) [reply hitman-go/Assets/Scripts/Enemy/EnemyService.cs]: removed it. previously i was using a dictionary. Now i am not

File: hitman-go/Assets/Scripts/Enemy/PatrollingEnemyController.cs (hunk 9 above)
Comment (VinayKashyap15) [reply hitman-go/Assets/Scripts/Enemy/PatrollingEnemyController.cs]: moved to view

File: hitman-go/Assets/Scripts/Enemy/IEnemyController.cs (hunk 10 above)
Comment (VinayKashyap15) [reply hitman-go/Assets/Scripts/Enemy/IEnemyController.cs]: resolved

File: hitman-go/Assets/Scripts/3StarSystem/IStarService.cs [hunk 18]
```diff
@@ -0,0 +1,11 @@
+using Common;
+namespace StarSystem
+{
+    public interface IStarService
+    {
+        void DogsKilled();
+        void AllDogsKilled();
+        bool CheckForStar(StarTypes starType);
+        void SetTotalEnemyandMaxPlayerMoves(int enemy, int playerMoves);
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/3StarSystem/IStarService.cs:9]: should be two separate functions SRP violation

File: hitman-go/Assets/Scripts/3StarSystem/StarSystemService.cs [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+{
+    public class StarSystemService : IStarService
+    {
+        bool noKill = true, briefCase = false, completed = false, killalldogs = false, killnodogs = true;
+        int totalEnemyInLevel, killCount = 0, playerMoves = 0, maxPlayerMoves;
+        SignalBus signalBus;
+        StarSystemService(SignalBus signalBus)
+        {
+            //PlayerPrefs.DeleteAll();
+            this.signalBus = signalBus;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/3StarSystem/StarSystemService.cs:17]: not even needed - not used outside the constructor

File: hitman-go/Assets/Scripts/3StarSystem/StarSystemService.cs [hunk 20]
```diff
@@ -0,0 +1,90 @@
+using Zenject;
+using UnityEngine;
+using Enemy;
+using GameState;
+using Common;
+using InteractableSystem;
+namespace StarSystem
+{
+    public class StarSystemService : IStarService
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/3StarSystem/StarSystemService.cs:9]: naming

File: hitman-go/Assets/Scripts/Camera/CameraManager.cs [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+using UnityEngine;
+using Common;
+using Zenject;
+using ScriptableObjSystem;
+using GameState;
+using PathSystem;
+
+namespace CameraSystem
+{
+    public class CameraManager : ICamera
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Camera/CameraManager.cs:12]: Manager or Service - be consistent

File: hitman-go/Assets/Scripts/Camera/CameraScriptable/CameraScriptableObj.cs [hunk 22]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace CameraSystem
+{
+    [System.Serializable]
+    public struct CameraData
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Camera/CameraScriptable/CameraScriptableObj.cs:8]: CameraModel

File: hitman-go/Assets/Scripts/Common/ScriptableNode.cs [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+
+namespace Common
+{
+    [Serializable]
+    public class ScriptableNode
+    {
+        
+        public NodeData node;
+        public int up = -1, down = -1, left = -1, right = -1;
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Common/ScriptableNode.cs:11]: why -1

File: hitman-go/Assets/Scripts/Editor/GraphValidator.cs [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        ScriptableGraph graph;
+        int[,] position_values;
+        int hzero, vzero;
+        public void OnEnable()
+        {
+            graph = (ScriptableGraph)target;
+            position_values = new int[graph.maxhight, graph.maxwidth];
+            int multiplier = 5;
+            hzero = ((int)(Mathf.Ceil(((float)graph.maxwidth / 2)))) - 1; vzero = ((int)(Mathf.Ceil(((float)graph.maxhight / 2))) - 1);
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Editor/GraphValidator.cs:20]: super hard to read

File: hitman-go/Assets/Scripts/Enemy/Controllers/BiDirectionalGuardEnemyController.cs [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+using UnityEngine;
+using Zenject;
+
+namespace Enemy
+{
+    public class BiDirectionalEnemyController : EnemyController
+    {
+        private Directions secondDirection;
+
+        public BiDirectionalEnemyController(IEnemyService _enemyService, IPathService _pathService, IGameService _gameService, Vector3 _spawnLocation, EnemyScriptableObject _enemyScriptableObject, int currentNodeID, Directions spawnDirection, bool _hasShield) : base(_enemyService, _pathService, _gameService, _spawnLocation, _enemyScriptableObject, currentNodeID, spawnDirection, _hasShield)
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/Controllers/BiDirectionalGuardEnemyController.cs:16]: is this readable?

File: hitman-go/Assets/Scripts/Enemy/Controllers/CircularCopEnemyController.cs [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            enemyType = EnemyType.CIRCULAR_COP;
+
+        }
+        protected override void SetController()
+        {
+            currentEnemyView.SetCurrentController(this);
+        }
+
+        async public override Task Move()
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/Controllers/CircularCopEnemyController.cs:23]: too big - single function
plus state machine logic

File: hitman-go/Assets/Scripts/Enemy/Controllers/EnemyController.cs [hunk 27]
```diff
@@ -0,0 +1,243 @@
...
+
+        }
+
+        protected virtual bool CheckForPlayerPresence(int _nextNodeID)
+        {
+            return (currentEnemyService.GetPlayerNodeID() == _nextNodeID);
+
+        }
+
+        protected virtual void ChangeDirection()
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/Controllers/EnemyController.cs:139]: duplicate logic from Dual facing cop

File: hitman-go/Assets/Scripts/Enemy/Controllers/EnemyController.cs [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+       public virtual bool IsKillable(KillMode killMode)
+        {
+            if (hasShield && killMode == KillMode.SHOOT)
+            {
+                return false;
+            }
+            else { return true; }
+        }
+
+       public bool IsPlayerKillable()
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/Controllers/EnemyController.cs:228]: why is this needed ?

File: hitman-go/Assets/Scripts/Enemy/Controllers/PatrollingEnemyController.cs [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+
+        public PatrollingEnemyController(IEnemyService _enemyService, IPathService _pathService, IGameService _gameService, Vector3 _spawnLocation, EnemyScriptableObject _enemyScriptableObject, int currentNodeID, Directions spawnDirection, bool _hasShield) : base(_enemyService, _pathService, _gameService, _spawnLocation, _enemyScriptableObject, currentNodeID, spawnDirection, _hasShield)
+        {
+            enemyType = EnemyType.PATROLLING;
+
+        }
+
+       async protected override Task MoveToNextNode(int nodeID)
+        {    
+            if (nodeID == -1)
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/Controllers/PatrollingEnemyController.cs:21]: no clue what -1 means

File: hitman-go/Assets/Scripts/Enemy/EnemyFactory.cs [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+        private List<IEnemyController> SpawnSingleEnemyLocations(EnemyScriptableObject _enemyScriptableObject)
+        {
+            
+
+            List<EnemySpawnData> spawnNodeID = new List<EnemySpawnData>();
+            List<IEnemyController> newEnemyControllers = new List<IEnemyController>();
+            spawnNodeID.Clear();
+            spawnNodeID = pathService.GetEnemySpawnLocation(_enemyScriptableObject.enemyType);
+
+            switch (_enemyScriptableObject.enemyType)
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/EnemyFactory.cs:54]: too much duplication in each switch

File: hitman-go/Assets/Scripts/Enemy/EnemyService.cs [hunk 31]
```diff
@@ -0,0 +1,258 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using StarSystem;
+using System.Threading.Tasks;
+using UnityEngine;
+using Zenject;
+
+namespace Enemy
+{
+    public class EnemyService : IEnemyService
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/EnemyService.cs:16]: 250 Loc

File: hitman-go/Assets/Scripts/Enemy/States/EnemyChaseState.cs [hunk 32]
```diff
@@ -0,0 +1,29 @@
+﻿using UnityEngine;
+using Common;
+using System.Collections;
+
+namespace Enemy
+{
+    public class EnemyChaseState : IEnemyState
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Enemy/States/EnemyChaseState.cs:7]: enemy states have no logic - they are just empty class

File: hitman-go/Assets/Scripts/InputSystem/Interfaces/ITapDetect.cs [hunk 33]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace InputSystem
+{
+    public interface ITapDetect
+    {
+        GameObject ReturnObject(Vector2 position, LayerMask layerMask);
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/InputSystem/Interfaces/ITapDetect.cs:9]: not a good name

File: hitman-go/Assets/Scripts/Player/PlayerService.cs [hunk 34]
```diff
@@ -0,0 +1,230 @@
...
+        {
+            signalBus.TryFire(new LevelFinishedSignal());
+        }
+
+        public void SetTargetTap(int tapNode)
+        {
+            targetNode = tapNode;
+        }
+
+        public SignalBus GetSignalBus()
```
Comment (mayankgrover) [inline hitman-go/Assets/Scripts/Player/PlayerService.cs:225]: why are these kind of functions needed?

**Review by mayankgrover — APPROVED** [review]

**Pull request merged**